### ISSUE TITLE
Stencil/color mask tool for locking colors

### ISSUE THREAD

**Issue by kpkpkpkpkpkpkpkp** [aseprite/aseprite#3855]
Pro Motion NG and old graphics applications like deluxe paint have a tool called 'Stencil' that lets you select any number of colors from the palette, and those colors will never be changed anywhere you're painting while the stencil mask is on. It's very flexible and straightforward way to lock transparent pixels for shading, or creating complex gradients and all kinds of other things. It'd be great to have in Aseprite. 

Selecting the color range you don't want to change and then inverting works basically the same, but it's time consuming to pick a lot of colors, not very performant on large images, and once you select you lose access to the selection tool for anything else. Also, with a stencil, if you paint using a color that is part of the stencil the new painted area becomes protected automatically, while using selection would require you to re-select the color.

I was looking for other feature requests for this king of thing and saw [#459 ](https://github.com/aseprite/aseprite/issues/459) but stencils are useful in addition to mask layers since with a stencil you don't need to create a new layer and paint the whole area that you don't want to cover, while mask layers are useful since they don't rely on color at all and are non-destructive.

The way I see it working is having some options in the palette options menu near 'edit palette'. One 'stencil mask' to toggle it, then 'edit stencil' would let you click on the colors in the palette you want to lock/unlock, or use eye-dropper to quickly select them from the current image, and finally an 'invert stencil' and 'clear stencil' option.

**Comment by Gasparoken** [comment]
Thanks for this suggestion. I like the stencil approach, anyway I feel that we can do almost the same thing adding the following features during 'Color Range' dialog:
- Capability for replace, adding and substract selection
- Capability for affect All Layers or Selected Layers
- Eyedropper tool by default to pick colors to select.
- Capability for zooming/panning
- Capability for save/load a color sets in .pal/.aseprite format (just for the selection purpose).

Mockup:
![SelectColorMockup](https://github.com/aseprite/aseprite/assets/32165475/8cbf2e94-abda-4ecb-9fda-31aa1be17cd0)

What do you think @dacap ?

**Comment by dacap** [comment]
The most similar features we have at the moment are the selection and the [shading ink](https://aseprite.org/docs/shading/) but both has the problem that we cannot transform the selection (and this is the main problem that cannot be solved easily).

Improving the Color Range dialog is an option and it'd be nice, but it doesn't solve this issue (as it's still have the problems that the selection cannot be transformed and cannot be re-selected quickly).

A plugin/script to select all colors from [app.range.colors](https://www.aseprite.org/api/range?l=Range&t=range#rangecolors) could be possible, even one that save/load different set of colors (it could solve the quick selection of several colors). But it still uses the selection (no "stencil" concept, but close enough).

I think this needs more thought and maybe the union with other feature (it sounds similar to Lock Alpha, like "Lock Colors"). I would wait to see if more users are interested in a similar feature, because at the moment Mask Layers are a more common concept in modern software.

**Comment by Gasparoken** [comment]
> ... it sounds similar to Lock Alpha, like "Lock Colors"...

I like this idea, an "Lock Color" ink where you can select the locked color range (in the same way as "Shading" ink). The current "Lock Alpha" would be a particular case of "Lock Alpha".

We will see the impressions of the artists.

**Comment by kpkpkpkpkpkpkpkp** [comment]
Yeah, pretty much exactly like lock alpha just for a range instead. And also right on it being similar to shading ink mode, just more flexible.

I really would like it to work independently of selection though. mostly because I want to be able to switch back and forth between frames and layers and whatever I want without having to go through a dropdown and a dialog to re-select the colors each time I switch. I guess the use cases are mostly the same as shading ink mode, just that I'd like to be able to use more colors more freely. So maybe it could just be another ink mode for the brush. I'd be pretty happy with that. It could even use the exact same UI elements from shading mode, just instead of a gradient it would be a list of the stencil colors.

It is kind of an old feature for sure, part of me thinks it just fell out of favor when graphics programs moved away from indexed color mode, and that people would use learn to love it if it were there. Maybe a "they don't know what they're missing" type of thing lol. Still understandable that it'd be a lower priority feature if it's just me asking for it.
